### ISSUE TITLE
"No processes were matched" after set parameter KINETO_DAEMON_INIT_DELAY_S

### ISSUE THREAD

**Issue by KepingYan** [facebookincubator/dynolog#286]
Test script: `scripts/pytorch/linear_model_example.py`
If I run it with command `KINETO_USE_DAEMON=1 python3 scripts/pytorch/linear_model_example.py`, I will meet this error as described in docs [manifests as a segfault](https://github.com/pytorch/pytorch/issues/131020).
Then I add parameter KINETO_DAEMON_INIT_DELAY_S to run this script `KINETO_USE_DAEMON=1 KINETO_DAEMON_INIT_DELAY_S=3 python3 scripts/pytorch/linear_model_example.py`,  but the trace file won't be generated as expected.
```
$ dyno gputrace --log-file /tmp/libkineto_trace.json
Kineto config =
ACTIVITIES_LOG_FILE=/tmp/libkineto_trace.json
PROFILE_START_TIME=0
ACTIVITIES_DURATION_MSECS=500
PROFILE_REPORT_INPUT_SHAPES=false
PROFILE_PROFILE_MEMORY=false
PROFILE_WITH_STACK=false
PROFILE_WITH_FLOPS=false
PROFILE_WITH_MODULES=false
response length = 133
response = {"activityProfilersBusy":0,"activityProfilersTriggered":[],"eventProfilersBusy":0,"eventProfilersTriggered":[],"processesMatched":[]}
No processes were matched, please check --job-id or --pids flags
```

Environment:
```
Ubuntu 22.04
torch     2.4.0
dynolog install:
wget https://github.com/facebookincubator/dynolog/releases/download/v0.3.2/dynolog_0.3.2-0-amd64.deb
sudo dpkg -i dynolog_0.3.2-0-amd64.deb
sudo systemctl restart dynolog
```

**Comment by briancoutinho** [comment]
Hi @KepingYan , do you see this print "INFO:2022-10-22 00:59:13 151209:151209 init.cpp:98] Registering daemon config loader" [docs](https://github.com/facebookincubator/dynolog/blob/main/docs/pytorch_profiler.md)
We can run dynolog, example script and dyno cmd in 3 different shells to get the expected behavior.

Also to help debug, does the machine have GPUs? Ideally, this should work for both CPU only and CPU+GPU systems but just checking

**Comment by htbig** [comment]
i have met the same thing,i have gpu,and i also see "INFO:2024-10-01 12:29:42 271363:271364 init.cpp:131] Registering daemon config loader, cpuOnly =  0",but i also see "response = {"activityProfilersBusy":0,"activityProfilersTriggered":[],"eventProfilersBusy":0,"eventProfilersTriggered":[],"processesMatched":[]}
No processes were matched, please check --job-id or --pids flags",i have point the --pids like this:sudo ./build/bin/dyno gputrace --log-file ./a.log --pids 271363

**Comment by htbig** [comment]
@KepingYan ,i have resolved this problem by add time.sleep(5) before i use GPU device in my script. you should have a try.

**Comment by KepingYan** [comment]
Hi @briancoutinho, this is the log of `python scripts/pytorch/linear_model_example.py` terminal, and it is executed on CPU+GPU(V100) systems.
```
$ KINETO_USE_DAEMON=1 KINETO_DAEMON_INIT_DELAY_S=1 python3 scripts/pytorch/linear_model_example.py
INFO:2024-10-09 14:45:19 384686:384687 init.cpp:131] Registering daemon config loader, cpuOnly =  0
ERROR: External init callback must run in same thread as registerClient (-932395456 != 2078357312)
cuda:0
99 874.3150024414062
10099 8.817167282104492
20099 8.817167282104492
30099 8.817168235778809
40099 8.817168235778809
50099 8.817168235778809
60099 8.817168235778809
70099 8.817168235778809
80099 8.817168235778809
90099 8.817168235778809
100099 8.817168235778809
110099 8.817168235778809
120099 8.817168235778809
130099 8.817168235778809
140099 8.817168235778809
150099 8.817168235778809
160099 8.817168235778809
170099 8.817168235778809
180099 8.817168235778809
190099 8.817168235778809
Result: y = 1.862860976586944e-09 + 0.8567266464233398 x + -1.0942008188408181e-08 x^2 + -0.09332837164402008 x^3
```
I've tried @htbig advice, but it doesn't work for my environment.

**Comment by zdddddda** [comment]
> Test script: `scripts/pytorch/linear_model_example.py` If I run it with command `KINETO_USE_DAEMON=1 python3 scripts/pytorch/linear_model_example.py`, I will meet this error as described in docs [manifests as a segfault](https://github.com/pytorch/pytorch/issues/131020). Then I add parameter KINETO_DAEMON_INIT_DELAY_S to run this script `KINETO_USE_DAEMON=1 KINETO_DAEMON_INIT_DELAY_S=3 python3 scripts/pytorch/linear_model_example.py`, but the trace file won't be generated as expected.
> 
> ```
> $ dyno gputrace --log-file /tmp/libkineto_trace.json
> Kineto config =
> ACTIVITIES_LOG_FILE=/tmp/libkineto_trace.json
> PROFILE_START_TIME=0
> ACTIVITIES_DURATION_MSECS=500
> PROFILE_REPORT_INPUT_SHAPES=false
> PROFILE_PROFILE_MEMORY=false
> PROFILE_WITH_STACK=false
> PROFILE_WITH_FLOPS=false
> PROFILE_WITH_MODULES=false
> response length = 133
> response = {"activityProfilersBusy":0,"activityProfilersTriggered":[],"eventProfilersBusy":0,"eventProfilersTriggered":[],"processesMatched":[]}
> No processes were matched, please check --job-id or --pids flags
> ```
> 
> Environment:
> 
> ```
> Ubuntu 22.04
> torch     2.4.0
> dynolog install:
> wget https://github.com/facebookincubator/dynolog/releases/download/v0.3.2/dynolog_0.3.2-0-amd64.deb
> sudo dpkg -i dynolog_0.3.2-0-amd64.deb
> sudo systemctl restart dynolog
> ```

I have the same question. How do you solve it?

**Comment by zdddddda** [comment]
Restart dynolog.service failed.  dynolog.service - Dynolog performance monitoring daemon
   Loaded: loaded (/lib/systemd/system/dynolog.service; enabled; vendor preset: enabled)
   Active: failed (Result: signal) since Thu 2025-01-16 22:16:03 CST; 4min 28s ago
  Process: 597807 ExecStartPre=/usr/bin/touch /etc/dynolog.gflags (code=exited, status=0/SUCCESS)
  Process: 597808 ExecStartPre=/usr/bin/touch /var/log/dynolog.log (code=exited, status=0/SUCCESS)
  Process: 597809 ExecStart=/usr/local/bin/dynolog --flagfile=/etc/dynolog.gflags (code=killed, signal=SEGV

**Comment by yinghai** [comment]
I had the same issue. 

client
```
KINETO_USE_DAEMON=1 KINETO_DAEMON_INIT_DELAY_S=3 python scripts/pytorch/linear_model_example.py
INFO:2025-01-29 01:45:27 43221:43222 init.cpp:135] Registering daemon config loader, cpuOnly =  0
ERROR: External init callback must run in same thread as registerClient (375600704 != -775577600)
cuda:0
99 316.04437255859375
...
```
then dyno
```
response = {"activityProfilersBusy":0,"activityProfilersTriggered":[],"eventProfilersBusy":0,"eventProfilersTriggered":[],"processesMatched":[]}
No processes were matched, please check --job-id or --pids flags
```

server side
```
./build/bin/dynolog --enable_ipc_monitor


I20250129 01:45:33.642140 43206 LibkinetoConfigManager.cpp:266] Initiating on-demand GPU profiling for job ID 0, pids [0]
I20250129 01:45:33.642149 43206 LibkinetoConfigManager.cpp:307] On-demand request: 0 matching processes
I20250129 01:45:33.642153 43206 LibkinetoConfigManager.cpp:315] Installed activity profiler config for 0 process(es) (0 busy)
```

**Comment by yinghai** [comment]
```
Initiating on-demand GPU profiling for job ID 0, pids [0]
```

this seems pretty weird?

**Comment by yinghai** [comment]
More info. When i start the script without `KINETO_DAEMON_INIT_DELAY_S=3`, on the dyno_log server side, I can see that the script PID is registered
```
I20250129 04:42:10.183240 43204 LibkinetoConfigManager.cpp:194] Registered process (43690, 25650, 24880, 1428) for job 0.
I20250129 04:42:10.183279 43204 IPCMonitor.cpp:84] getLibkinetoOnDemandRequest() : job id 0 pids = 43690
```

when I start the script with `KINETO_DAEMON_INIT_DELAY_S=3`, however, there is no log on the dyno_log server side which seems to suggest that the pytorch program is not sending the message to dyno_log server?

**Comment by yinghai** [comment]
On the client side, when I don't use `KINETO_DAEMON_INIT_DELAY_S=3`, I see a bunch of kineto logs being printed out 
```
INFO:2025-01-29 05:13:33 44126:44126 init.cpp:135] Registering daemon config loader, cpuOnly =  1
INFO:2025-01-29 05:13:36 44126:44126 CuptiActivityProfiler.cpp:241] CUDA versions. CUPTI: 22; Runtime: 12040; Driver: 12040
INFO:2025-01-29 05:13:36 44126:44126 DaemonConfigLoader.cpp:64] Setting communication fabric enabled = 1
INFO:2025-01-29 05:13:36 44126:44128 DaemonConfigLoader.cpp:64] Setting communication fabric enabled = 1
INFO:2025-01-29 05:13:36 44126:44126 IpcFabricConfigClient.cpp:88] Setting up IPC Fabric at endpoint: dynoconfigclientc45743f1-92bd-46e5-a6fa-2e4d1a9484fe stat
us = initialized
```
which makes sense as then it will send message to dynolog server
https://github.com/pytorch/kineto/blob/09e10ad38fdcdcba065f2e5871a219fa9a9f0067/libkineto/src/IpcFabricConfigClient.cpp#L151-L175

but when I use `KINETO_DAEMON_INIT_DELAY_S=3`, I only got 
```
INFO:2025-01-29 05:25:14 44304:44305 init.cpp:135] Registering daemon config loader, cpuOnly =  0
ERROR: External init callback must run in same thread as registerClient (1258501696 != 107225088)
```

no log like `INFO:2025-01-29 05:13:36 44126:44126 IpcFabricConfigClient.cpp:88] Setting up IPC Fabric at endpoint: dynoconfigclientc45743f1-92bd-46e5-a6fa-2e4d1a9484fe stat`

I wonder if 
```
ERROR: External init callback must run in same thread as registerClient (1258501696 != 107225088)
```
is legit error that is preventing kineto init from making further progress?

**Comment by yinghai** [comment]
Update 
I was able to build pytorch/kineto and play around, so far I think 
```
ERROR: External init callback must run in same thread as registerClient (1258501696 != 107225088)
```
is not related as even if I patch https://github.com/pytorch/pytorch/commit/c934ed65673428c35f70cbce6ebd730ad22f058d and make the above error go away, the issue still exists. 

So far I was able to localize the issue being 
https://github.com/pytorch/kineto/blob/d9753139d181b9ff42872465aac0e5d3018be415/libkineto/src/ActivityProfilerProxy.cpp#L27 not being called when cpu_only=0, while it is called when cpu_only=1. 

@briancoutinho do you know who how the init() function above is triggered?

**Comment by yinghai** [comment]
traced a bit more, I think the difference is when cpu_only=1, we hit this code path
https://github.com/pytorch/kineto/blob/0cded49b0cce22f7d777861db5c937da7f3e22db/libkineto/src/init.cpp#L221
which triggers the `libkineto::api().initProfilerIfRegistered();` call. 

When cpu_only=0, supposingly we should hit this function https://github.com/pytorch/kineto/blob/0cded49b0cce22f7d777861db5c937da7f3e22db/libkineto/src/init.cpp#L62 upon first cuda context creation: https://github.com/pytorch/kineto/blob/0cded49b0cce22f7d777861db5c937da7f3e22db/libkineto/src/init.cpp#L153-L154 but looks like this is not being triggered. Will check why.

**Comment by yinghai** [comment]
these operations are all successful
https://github.com/pytorch/kineto/blob/0cded49b0cce22f7d777861db5c937da7f3e22db/libkineto/src/init.cpp#L146-L156

I guess somehow the cupti `RESOURCE_CONTEXT_CREATED` hook is not reliable anymore. 

my cuda info
```
Driver Version: 550.144.03     CUDA Version: 12.4
```

**Comment by briancoutinho** [comment]
Hi @yinghai , just wanted to confirm, the original issue is caused by 
"External init callback must run in same thread as registerClient"

The issue is when we add the delay, PyTorch uses a different thread to init the profiler state, and this was clashing with this check. It used to work before but it started failing here at some point in PyTorch-
https://github.com/pytorch/kineto/blob/d9753139d181b9ff42872465aac0e5d3018be415/libkineto/src/libkineto_api.cpp#L23-L31

## Fix
In a more recent versions of PyTorch, you don't need the delay anymore and the init function is called when we "import torch". You shared the patch above too
https://github.com/pytorch/pytorch/pull/131448

## Init Flow explanation
For completeness, just wanted to share the init  flow 
1. With above patch, when torch is loaded it will cal libkineto_init from "global_kineto_init()" function. Cpu only will be set to 0 assuming a CUDA build.
2. For actual init
2a  If it is a CPU build the init is done immediately.  
https://github.com/pytorch/kineto/blob/d9753139d181b9ff42872465aac0e5d3018be415/libkineto/src/init.cpp#L220-L221
2.b For a CUDA build, the init is registered as a callback on the creation of a CUDA Context
https://github.com/pytorch/kineto/blob/d9753139d181b9ff42872465aac0e5d3018be415/libkineto/src/init.cpp#L142-L158
The callback function is "initProfilers()".
3. Either options above will call the underlying functions in ActivityProfiler* files ->  that eventually register with dynolog.

I think callbacks were not triggered, could you check if you run "nvidia-smi" that the process id shows up? Then it means there is a valid CUDA context.

**Comment by briancoutinho** [comment]
Catching up on other messages. I am not sure why we are not triggering on CUDA context creation.

With your patched build we should not be delaying the init as well. Just to confirm, can you remove the delay variable too
```
KINETO_USE_DAEMON=1 python scripts/pytorch/linear_model_example.py
```

**Comment by yinghai** [comment]
So yeah, the KINETO_DAEMON_INIT_DELAY_S env is actually not the root cause. I have patched the fix https://github.com/pytorch/pytorch/commit/c934ed65673428c35f70cbce6ebd730ad22f058d and get rid of KINETO_DAEMON_INIT_DELAY_S but the issue remains. 

I think all it does is to make sure that we are doing cpu_only=0 mode and init by doing the callback and then the same issue persists which is the context creation callback doesn't trigger. I'll check nvidia-smi as you suggested shortly. 

One thing I don't fully understand on this lazy loading cupti thing though. I guess the purpose of doing callback is to defer cupti init but according the cupti document, cupti lib is loaded whenever the first call to cupti API is triggered, but isn't the cuptiApiCallback function also a call to cupti API?

**Comment by yinghai** [comment]
`nvidia-smi` shows processed but it's possibly that I'm running in a container (https://github.com/NVIDIA/nvidia-docker/issues/179). Since scripts/pytorch/linear_model_example.py is running I think the context should be there already.

**Comment by yinghai** [comment]
The part of dynamic attach in cupti: https://docs.nvidia.com/cupti/main/main.html#dynamic-attach-and-detach 

> CUPTI can be attached by calling any CUPTI API as CUPTI supports lazy initialization.

so when we do something like `CuptiCallbackApi::initCallbackApi()` which calls `cuptiSubscribe`, it already attaches the cupti which sort of already defeat the purpose of lazy loading?

**Comment by briancoutinho** [comment]
@yinghai The intent with that logic was mainly to delay the initialization to a later point. It didn't have to do with CUPTI itself. This is historical reason.
With the new change that enables this on "import torch" i am wondering if we should just do initialization here instead of enqueue in the callback.

Since you have the local build, could you try removing the "cpuOnly" condition from this if statement?
https://github.com/pytorch/kineto/blob/d9753139d181b9ff42872465aac0e5d3018be415/libkineto/src/init.cpp#L220

That should work i believe?

**Comment by yinghai** [comment]
yeah exactly, I actually already got it working last night by doing 
```
initProfilersCPU();
    libkineto::api().configLoader().initBaseConfig();
```
regardless of cpu_only. I think it is a legit change as 
- all the EventProfiler setup is no longer recommended since CUDA 10.0. 
- it doesn't regrade the behavior in terms of laziness as I think when we start a cuda program we likely will need to create cuda context very soon. 

But there is another problem, unrelated to the issue here. I think the laziness here might not be working as we wanted as in `global_kineto_init` we are bringing cupti stuff immediately by calling any cupti functions (which could be as simple as `cuptiGetVersion`). Then we are potentially affecting the perf of the program until `cuptiFinalize` is called. Which seems to be an issue?

**Comment by yinghai** [comment]
Yeah makes sense! Thanks a lot!

**Comment by yinghai** [comment]
I created a separate issue in libkineto repo: https://github.com/pytorch/kineto/issues/1032

**Comment by briancoutinho** [comment]
@lshmouse  I posted some context up here https://github.com/facebookincubator/dynolog/issues/286#issuecomment-2622171475

But the issue is not fully fixed. I am working on a fix https://github.com/pytorch/kineto/issues/1032 and hopefully changes should be available in pytorch nightly sometime coming week

follow along in this issue https://github.com/pytorch/kineto/issues/1032

**Comment by lshmouse** [comment]
> [@lshmouse](https://github.com/lshmouse) I posted some context up here [#286 (comment)](https://github.com/facebookincubator/dynolog/issues/286#issuecomment-2622171475)
> 
> But the issue is not fully fixed. I am working on a fix [pytorch/kineto#1032](https://github.com/pytorch/kineto/issues/1032) and hopefully changes should be available in pytorch nightly sometime coming week
> 
> follow along in this issue [pytorch/kineto#1032](https://github.com/pytorch/kineto/issues/1032)

Thanks~

**Comment by lshmouse** [comment]
I have tested many old pytorch versions:  the issue is started with pytorch 2.3 and python <=2.2.2  don't have this issue.

**Comment by briancoutinho** [comment]
@lshmouse thanks for sharing the versions that work :) 

If you move to the pytorch nightly past Feb 15th, 2025 the issue should now be fixed. This should also show up in pytorch 2.2.7 once the release is made. Will keep folks posted and update the documentation.
